ocs_ci/deployment/provider_client/storage_client_deployment.py

```diff
@@ -0,0 +1,550 @@
+        )
+        self.storage_profile_obj = ocp.OCP(
+            kind="Storageprofile", namespace=config.ENV_DATA["cluster_namespace"]
+        )
+        self.config_map_obj = ocp.OCP(
+            kind="Configmap", namespace=config.ENV_DATA["cluster_namespace"]
+        )
+        self.pod_obj = ocp.OCP(
+            kind="Pod", namespace=config.ENV_DATA["cluster_namespace"]
+        )
+        self.scheduler_obj = ocp.OCP(
+            kind=constants.SCHEDULERS_CONFIG,
+            namespace=config.ENV_DATA["cluster_namespace"],
+        )
+        self.sc_obj = ocp.OCP(kind=constants.STORAGECLASS)
+        self.storageclass = "localblock"
+        self.ocp_version = version.get_semantic_ocp_version_from_config()
+        self.ocs_version = version.get_semantic_ocs_version_from_config()
+        self.storage_class_claims = [
+            constants.CEPHBLOCKPOOL_SC,
+            constants.CEPHFILESYSTEM_SC,
+        ]
+        self.ocs_client_operator = defaults.OCS_CLIENT_OPERATOR_NAME
+        self.deployment = Deployment()
+        self.rados_utils = RadosHelper()
+
+    def provider_and_native_client_installation(
+        self,
+    ):
+        """
+        1. set control nodes as scheduleable
+        2. allow ODF to be deployed on all nodes
+        3. allow hosting cluster domain to be usable by hosted clusters
+        4. Enable nested virtualization on vSphere nodes
+        5. Install ODF
+        6. Install LSO, create LocalVolumeDiscovery and LocalVolumeSet
+        7. Disable ROOK_CSI_ENABLE_CEPHFS and ROOK_CSI_ENABLE_RBD
+        8. Create storage profile
+        """
+
+        if (
+            self.ocs_version < version.VERSION_4_16
+            and self.ocs_version >= version.VERSION_4_14
+        ):
+            # set control nodes as scheduleable
+            path = "/spec/mastersSchedulable"
+            params = f"""[{{"op": "replace", "path": "{path}", "value": true}}]"""
+            self.scheduler_obj.patch(params=params, format_type="json"), (
+                "Failed to run patch command to update control nodes as scheduleable"
+            )
+
+        # Allow ODF to be deployed on all nodes
+        nodes = get_all_nodes()
+        node_objs = get_node_objs(nodes)
+
+        log.info("labeling storage nodes")
+        label_nodes(nodes=node_objs, label=constants.OPERATOR_NODE_LABEL)
+
+        # Allow hosting cluster domain to be usable by hosted clusters
+        path = "/spec/routeAdmission"
+        value = '{wildcardPolicy: "WildcardsAllowed"}'
+        params = f"""[{{"op": "add", "path": "{path}", "value": {value}}}]"""
+        patch_cmd = (
+            f"patch {constants.INGRESSCONTROLLER} -n {constants.OPENSHIFT_INGRESS_OPERATOR_NAMESPACE} "
+            + f"default --type json -p '{params}'"
+        )
+        self.ocp_obj.exec_oc_cmd(command=patch_cmd)
+
+        # Enable nested virtualization on nodes
+        machine_config_data = templating.load_yaml(constants.MACHINE_CONFIG_YAML)
```

Comment:
The machine config yaml raises an error:
```
2024-05-30 13:34:55  ______________________ ERROR at setup of test_deployment _______________________
2024-05-30 13:34:55  
2024-05-30 13:34:55  request = <SubRequest 'cluster' for <Function test_deployment>>
2024-05-30 13:34:55  log_cli_level = 'INFO'
2024-05-30 13:34:55  record_testsuite_property = <bound method LogXML.add_global_property of <_pytest.junitxml.LogXML object at 0x7f81fafd5ac0>>
2024-05-30 13:34:55  set_live_must_gather_images = None
2024-05-30 13:34:55  
2024-05-30 13:34:55      @pytest.fixture(scope="session", autouse=True)
2024-05-30 13:34:55      def cluster(
2024-05-30 13:34:55          request, log_cli_level, record_testsuite_property, set_live_must_gather_images
2024-05-30 13:34:55      ):
2024-05-30 13:34:55          """
2024-05-30 13:34:55          This fixture initiates deployment for both OCP and OCS clusters.
2024-05-30 13:34:55          Specific platform deployment classes will handle the fine details
2024-05-30 13:34:55          of action
2024-05-30 13:34:55          """
2024-05-30 13:34:55          log.info(f"All logs located at {ocsci_log_path()}")
2024-05-30 13:34:55      
2024-05-30 13:34:55          teardown = ocsci_config.RUN["cli_params"]["teardown"]
2024-05-30 13:34:55          deploy = ocsci_config.RUN["cli_params"]["deploy"]
2024-05-30 13:34:55          if teardown or deploy:
2024-05-30 13:34:55              factory = dep_factory.DeploymentFactory()
2024-05-30 13:34:55              deployer = factory.get_deployment()
2024-05-30 13:34:55      
2024-05-30 13:34:55          # Add a finalizer to teardown the cluster after test execution is finished
2024-05-30 13:34:55          if teardown:
2024-05-30 13:34:55      
2024-05-30 13:34:55              def cluster_teardown_finalizer():
2024-05-30 13:34:55                  # If KMS is configured, clean up the backend resources
2024-05-30 13:34:55                  # we are doing it before OCP cleanup
2024-05-30 13:34:55                  if ocsci_config.DEPLOYMENT.get("kms_deployment"):
2024-05-30 13:34:55                      try:
2024-05-30 13:34:55                          kms = KMS.get_kms_deployment()
2024-05-30 13:34:55                          kms.cleanup()
2024-05-30 13:34:55                      except Exception as ex:
2024-05-30 13:34:55                          log.error(f"Failed to cleanup KMS. Exception is: {ex}")
2024-05-30 13:34:55                  deployer.destroy_cluster(log_cli_level)
2024-05-30 13:34:55      
2024-05-30 13:34:55              request.addfinalizer(cluster_teardown_finalizer)
2024-05-30 13:34:55              log.info("Will teardown cluster because --teardown was provided")
2024-05-30 13:34:55      
2024-05-30 13:34:55          # Download client
2024-05-30 13:34:55          if ocsci_config.DEPLOYMENT["skip_download_client"]:
2024-05-30 13:34:55              log.info("Skipping client download")
2024-05-30 13:34:55          else:
2024-05-30 13:34:55              force_download = (
2024-05-30 13:34:55                  ocsci_config.RUN["cli_params"].get("deploy")
2024-05-30 13:34:55                  and ocsci_config.DEPLOYMENT["force_download_client"]
2024-05-30 13:34:55              )
2024-05-30 13:34:55              get_openshift_client(force_download=force_download)
2024-05-30 13:34:55      
2024-05-30 13:34:55          # set environment variable for early testing of RHCOS
2024-05-30 13:34:55          if ocsci_config.ENV_DATA.get("early_testing"):
2024-05-30 13:34:55              release_img = ocsci_config.ENV_DATA["RELEASE_IMG"]
2024-05-30 13:34:55              log.info(f"Running early testing of RHCOS with release image: {release_img}")
2024-05-30 13:34:55              os.environ["RELEASE_IMG"] = release_img
2024-05-30 13:34:55              os.environ["OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE"] = release_img
2024-05-30 13:34:55      
2024-05-30 13:34:55          if deploy:
2024-05-30 13:34:55              # Deploy cluster
2024-05-30 13:34:55  >           deployer.deploy_cluster(log_cli_level)
2024-05-30 13:34:55  
2024-05-30 13:34:55  tests/conftest.py:1788: 
2024-05-30 13:34:55  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-05-30 13:34:55  ocs_ci/deployment/deployment.py:605: in deploy_cluster
2024-05-30 13:34:55      self.do_deploy_odf_provider_mode()
2024-05-30 13:34:55  ocs_ci/deployment/deployment.py:542: in do_deploy_odf_provider_mode
2024-05-30 13:34:55      storage_client_deployment_obj.provider_and_native_client_installation()
2024-05-30 13:34:55  ocs_ci/deployment/provider_client/storage_client_deployment.py:153: in provider_and_native_client_installation
2024-05-30 13:34:55      machine_config_data = templating.load_yaml(constants.MACHINE_CONFIG_YAML)
2024-05-30 13:34:55  ocs_ci/utility/templating.py:159: in load_yaml
2024-05-30 13:34:55      return loader(fs.read())
2024-05-30 13:34:55  venv/lib64/python3.9/site-packages/yaml/__init__.py:125: in safe_load
2024-05-30 13:34:55      return load(stream, SafeLoader)
2024-05-30 13:34:55  venv/lib64/python3.9/site-packages/yaml/__init__.py:81: in load
2024-05-30 13:34:55      return loader.get_single_data()
2024-05-30 13:34:55  venv/lib64/python3.9/site-packages/yaml/constructor.py:49: in get_single_data
2024-05-30 13:34:55      node = self.get_single_node()
2024-05-30 13:34:55  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-05-30 13:34:55  
2024-05-30 13:34:55  self = <yaml.loader.SafeLoader object at 0x7f81f1772850>
2024-05-30 13:34:55  
2024-05-30 13:34:55      def get_single_node(self):
2024-05-30 13:34:55          # Drop the STREAM-START event.
2024-05-30 13:34:55          self.get_event()
2024-05-30 13:34:55      
2024-05-30 13:34:55          # Compose a document if the stream is not empty.
2024-05-30 13:34:55          document = None
2024-05-30 13:34:55          if not self.check_event(StreamEndEvent):
2024-05-30 13:34:55              document = self.compose_document()
2024-05-30 13:34:55      
2024-05-30 13:34:55          # Ensure that the stream contains no more documents.
2024-05-30 13:34:55          if not self.check_event(StreamEndEvent):
2024-05-30 13:34:55              event = self.get_event()
2024-05-30 13:34:55  >           raise ComposerError("expected a single document in the stream",
2024-05-30 13:34:55                      document.start_mark, "but found another document",
2024-05-30 13:34:55                      event.start_mark)
2024-05-30 13:34:55  E           yaml.composer.ComposerError: expected a single document in the stream
2024-05-30 13:34:55  E             in "<unicode string>", line 2, column 1:
2024-05-30 13:34:55  E               apiVersion: machineconfiguration ... 
2024-05-30 13:34:55  E               ^
2024-05-30 13:34:55  E           but found another document
2024-05-30 13:34:55  E             in "<unicode string>", line 21, column 1:
2024-05-30 13:34:55  E               ---
2024-05-30 13:34:55  E               ^
2024-05-30 13:34:55  
2024-05-30 13:34:55  venv/lib64/python3.9/site-packages/yaml/composer.py:41: ComposerError
```

Reply:
Addressed in latest commit.